velox/functions/sparksql/Comparisons.cpp

```diff
@@ -13,16 +13,91 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+#include "velox/functions/sparksql/LeastGreatest.h"
 
-#include <utility>
-
-#include "velox/expression/DecodedArgs.h"
-#include "velox/expression/VectorFunction.h"
+#include "velox/expression/EvalCtx.h"
+#include "velox/expression/Expr.h"
 #include "velox/functions/sparksql/Comparisons.h"
+#include "velox/type/Type.h"
 
 namespace facebook::velox::functions::sparksql {
 namespace {
 
+template <typename Cmp, TypeKind kind>
+class ComparisonFunction final : public exec::VectorFunction {
+  using T = typename TypeTraits<kind>::NativeType;
+
+  bool supportsFlatNoNullsFastPath() const override {
+    return true;
+  }
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    context.ensureWritable(rows, BOOLEAN(), result);
+    auto* flatResult = result->asFlatVector<bool>();
+    const Cmp cmp;
+
+    exec::DecodedArgs decodedArgs(rows, args, context);
+    auto decodedArg0 = decodedArgs.at(0);
+    auto decodedArg1 = decodedArgs.at(1);
+    if (decodedArg0->isIdentityMapping() && decodedArg1->isConstantMapping()) {
+      rows.applyToSelected([&](auto i) {
+        flatResult->set(
+            i, cmp(decodedArg0->valueAt<T>(i), decodedArg1->valueAt<T>(0)));
+      });
+    } else if (
+        decodedArg0->isConstantMapping() && decodedArg1->isIdentityMapping()) {
+      rows.applyToSelected([&](auto i) {
+        flatResult->set(
+            i, cmp(decodedArg0->valueAt<T>(0), decodedArg1->valueAt<T>(i)));
+      });
+    } else {
+      rows.applyToSelected([&](auto i) {
+        flatResult->set(
+            i, cmp(decodedArg0->valueAt<T>(i), decodedArg1->valueAt<T>(i)));
```

Comment:
When both the args are constants or identity mapping we can not use valueAt and compare buffers - presto uses simd to compare(https://github.com/facebookincubator/velox/blob/main/velox/functions/prestosql/Comparisons.cpp) but atleast bypassing valueAt would be a good start.

Reply:
Updated. Please take a look again. Thanks.

Reply:
@kgpai Have time to take a further look? There is a function signature mismatch in CI but I think it's expected?

Reply:
Will have a look soon !